implement: Keep Chat scroll position when leaving and re-entering it #471

sorewanya opened the pull request:
##Background
Messenger [Gapopa](https://gapopa.net/). Chat page. Scroll in chat.

##Problem to solve
To do: save scroll position when re-entering chat.

Suggested time to implement: ~2 days.

For example: open chat and scroll, after that go to any otherr page or a different chat, then go back to the chat and make sure that the scroll stays exactly the same as it was when leaving the chat.

##Requirements:

the scroll position should be saved when re-entering chat.
when closing the application the scroll position should NOT be saved.
Possible solutions
As a possibility, you can save the ID and offset of the top/bottom element in chat and reuse it. There should be no persist, since remembering the position is not required.


## Solution

Save it in the 'static Map' with the key id.val , and in the value the object with the selected object ChatItemId and the offset taken from _topVisibleItem


## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains issue reference
    - [x] Has type and `k::` labels applied
- Before [review][l:4]:
    - [x] Documentation is updated (if required)
    - [x] Tests are updated (if required)
    - [x] Changes conform [code style][l:2]
    - [x] [CHANGELOG entry][l:3] is added (if required)
    - [x] FCM (final commit message) is posted or updated
    - [ ] [Draft mode][l:1] is removed
- [ ] [Review][l:4] is completed and changes are approved
    - [ ] FCM (final commit message) is approved
- Before merge:
    - [ ] Milestone is set
    - [ ] PR's name and description are correct and up-to-date
    - [ ] All temporary labels are removed




[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: https://github.com/team113/messenger/blob/main/CONTRIBUTING.md#code-style
[l:3]: https://github.com/team113/messenger/blob/main/CHANGELOG.md
[l:4]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

sorewanya committed:
v1 tests

sorewanya committed:
v2

sorewanya committed:
v3

sorewanya committed:
minor changes

SleepySquash commented:
@sorewanya, this sounds like a bug, I guess.

sorewanya committed:
move store Map to ChatService

sorewanya committed:
just for history of changes: remove forgotten static

sorewanya committed:
e2e test v1

sorewanya committed:
move to RxChat

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'KeepChatScrollPosition' of https://github.com/sorewanya/messenger into KeepChatScrollPosition

sorewanya committed:
fix https://github.com/team113/messenger/pull/1484#issuecomment-3468936445 - use animateToIndex;
"fix" e2e test

sorewanya committed:
CHANGELOG, remove override'd doc

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

SleepySquash reviewed (changes requested):
Thank you for your work! Please, take a detailed look at the UX moments I've commented you about - those are essential for frontend application. Those things are first priority always.

Also please don't forget about Effective Dart and Contribution Guide. Those documents exist for the teams to be able to cooperate as efficiently as possible. And the review process shouldn't be considered as an opportunity for someone to comment you about those documents being violated. You should do a self-review ideally with those things in mind before sending the PR for a final review, please.

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
corrections:
use addPostFrameCallback;
don't animate, use jumpToIndex;
don't save index, use itemId

sorewanya committed:
Merge branch 'KeepChatScrollPosition' of https://github.com/sorewanya/messenger into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
deleted the unnecessary if

sorewanya committed:
Merge branch 'KeepChatScrollPosition' of https://github.com/sorewanya/messenger into KeepChatScrollPosition

SleepySquash commented on lib/domain/repository/chat.dart:
```diff
@@ -388,3 +397,14 @@ class RxChatMember implements Comparable<RxChatMember> {
 }
+
+/// A scroll position anchored to a specific chat item and its offset.
+class ChatScrollPosition {
```
Did you take a look at contribution guide after the last review request to do so?

https://github.com/team113/messenger/blob/main/CONTRIBUTING.md#documentation-1

> Documentation must follow [Effective Dart](https://dart.dev/guides/language/effective-dart) official recommendations with the following exception:
>
> - prefer omitting leading A, An or The article.

SleepySquash commented on lib/domain/repository/chat.dart:
I still can see sadly some jumping around in a chat with single message (it contains only images)

https://github.com/user-attachments/assets/0493886e-6282-4b95-8fee-4a28db33a01d

SleepySquash committed:
Minor corrections

SleepySquash reviewed (changes requested):
I noticed a strange frame still present during testing, sadly. I've attached a video demonstration to one of the comments. Take a look, please.

sorewanya commented on lib/domain/repository/chat.dart:
```diff
@@ -388,3 +397,14 @@ class RxChatMember implements Comparable<RxChatMember> {
+
+/// A scroll position anchored to a specific chat item and its offset.
+class ChatScrollPosition {
+  ChatScrollPosition({required this.itemId, required this.offset});
```
Sorry, I restored the lines to their original state several times while testing ideas for fixes, without rechecking the last time. Apparently, I won't be closing "Resolve conversation" until I send it back for a second review.

sorewanya committed:
lost const

sorewanya committed:
fix e2e test: remove And after Then

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya reviewed (commented):
for send pended

sorewanya committed:
animated variant

sorewanya committed:
add forceToExecuteInitIndex to index==0

sorewanya committed:
fix log

sorewanya committed:
don't use forceToExecuteInitIndex if offset not restored

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
fix sort

sorewanya committed:
add final's

sorewanya committed:
try fix scroll over-jump when animating bottom

sorewanya committed:
use _updateFabStates, not history

sorewanya committed:
variant to fix over-jump's

sorewanya committed:
fix doc

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

SleepySquash commented on lib/ui/page/home/page/chat/view.dart:
```diff
@@ -488,6 +488,10 @@ class ChatView extends StatelessWidget {
+                              forceToExecuteInitIndex:
+                                  c.initIndex == 0 && c.initOffset != 0
+                                  ? 0
+                                  : null,
```
Would you be so kind please to write a comment down here for the developers to understand what this thing does here? Time will pass and people will forget.

SleepySquash commented on lib/store/chat_rx.dart:
```diff
@@ -335,6 +338,25 @@ class RxChatImpl extends RxChat {
+      '$runtimeType',
+    );
+    return _scrollPosition;
+  }
```
If all that this `get`+`set` does is just setting/reading the variable itself, then is there a reason to do a `get`+`set` at all? Can't just a simple public variable be sufficient enough instead of doing two separate methods?

sorewanya replied:
inertia of thinking. Repeated abstract class declarations 1-on-1. There it is justified, here it is not.

SleepySquash reviewed (commented):
Some thoughts I had by looking at the code, please. Take a look, please!

sorewanya commented on lib/ui/page/home/page/chat/controller.dart:
```diff
@@ -2432,6 +2439,74 @@ class ChatController extends GetxController {
+    try {
+      final ChatItemId itemId = chat!.scrollPosition!.itemId;
+
+      await _fetchItemsAround(itemId);
```
This is an 'artifact' of the previous use of `jumpTo`. Considering that I removed it and use `initIndex` directly, it is really not needed.

sorewanya committed:
remove _fetchItemsAround

sorewanya committed:
doc

sorewanya committed:
clean ChatRxImpl

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
remove unneeded try

sorewanya committed:
Merge branch 'KeepChatScrollPosition' of https://github.com/sorewanya/messenger into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition

sorewanya committed:
Merge branch 'main' into KeepChatScrollPosition
